OrbitVulkanLayer/SubmissionTracker.h

```diff
@@ -0,0 +1,779 @@
+// So, we will use `QueueSubmission` as a return value, and therefore these structs can't be
+// internal to the `SubmissionTracker`.
+namespace internal {
+
+// Stores meta information about a submit (VkQueueSubmit), e.g. to allow to identify the matching
+// Gpu tracepoint.
+struct SubmissionMetaInformation {
+  uint64_t pre_submission_cpu_timestamp;
+  uint64_t post_submission_cpu_timestamp;
+  int32_t thread_id;
+};
+
+// A persistent version of a command buffer that was submitted and its begin/end slot in the
+// `TimerQueryPool`. Note that the begin is optional, as it might not be part of the capture.
+// This struct is created if we capture the submission, so if we haven't captured the end,
+// we also haven't captured the begin, and don't need to store info about that command buffer at
+// all. The list of all `SubmittedCommandBuffer`s is stored in a `QueueSubmission`, and can be so
+// associated to the `SubmissionMetaInformation`.
+struct SubmittedCommandBuffer {
+  std::optional<uint32_t> command_buffer_begin_slot_index;
+  uint32_t command_buffer_end_slot_index;
+};
+
+// A persistent version of a debug marker (either begin or end) that was submitted and its slot
+// in the `TimerQueryPool`. `SubmittedMarkers` are used in `MarkerState` to identify the begin or
+// end marker. All markers (`MarkerState`s) that gets *completed* within a certain submission are
+// stored in the `QueueSubmission`.
+// Note that we also store the `SubmissionMetaInformation` as begin markers
+// can originate in different submissions then the matching end markers. This allows us e.g. to
+// match the markers to a specific submission tracepoint. So even though the submitted markers a
+struct SubmittedMarker {
+  SubmissionMetaInformation meta_information;
+  uint32_t slot_index;
+};
+
+// Represents a color to be used to in debug markers. The values are all in range [0.f, 1.f].
+struct Color {
+  float red;
+  float green;
+  float blue;
+  float alpha;
+};
+
+// Identifies a particular debug marker region and has two purposes.
+// 1. We have a stack of all markers of a queue, that gets updated upon a submission
+// (VkQueueSubmit). So on a submitted "begin" marker, we create that struct and push it to the
+// stack. On a end, we pop from that stack and if we are capturing complete the information and move
+// it the completed markers of a `QueueSubmission`.
+// 2. Is the list of completed markers in `QueueSubmission` that makes the marker information
+// persistent, so that we can try to read the timestamps on a present.
+// Note that we only store the state into `QueueSubmission`, if at that time we have a value for the
+// end_info. So once it is submitted, the end_info will always be set.
+// Beside the information about the begin/end, it also stores the text, color and the depth of the
+// marker. If a debug marker begin was discarded because of its depth, cut_off is set to true.
```

Comment:
So we keep it on the stack but then we don't persist it? I'm still a bit confused.

Reply:
Exactly. The idea is, that we still need the information that there was this begin marker, in order to have the begins and ends being in balance/matched.